src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,713 @@
+(* SPDX-License-Identifier: AGPL-3.0-or-later *)
+(* Copyright © 2021-2024 OCamlPro *)
+(* Written by the Owi programmers *)
+
+open Binary
+open Syntax
+open Types
+
+let end_opcode = '\x0B'
+
+let magic = "\x00\x61\x73\x6d"
+
+let version = "\x01\x00\x00\x00"
+
+let add_char c buf = Buffer.add_char buf c
+
+let add_string str buf = Buffer.add_string buf str
+
+(* add byte from int (ascii code) *)
+let add_byte i =
+  let c = Char.chr (i land 0xff) in
+  add_char c
+
+(* add 2 bytes (16 bits) from int *)
+let add_bytes_2 i buf =
+  add_byte (i land 0xff) buf;
+  add_byte (i lsr 8) buf
+
+(* add 4 bytes (32 bits) from int32 *)
+let add_bytes_4 i buf =
+  add_bytes_2 (Int32.to_int (Int32.logand i 0xffffl)) buf;
+  add_bytes_2 (Int32.to_int (Int32.shift_right_logical i 16)) buf
+
+(* add 8 bytes (64 bits) from int64 *)
+let add_bytes_8 i buf =
+  add_bytes_4 (Int64.to_int32 (Int64.logand i 0xffffffffL)) buf;
+  add_bytes_4 (Int64.to_int32 (Int64.shift_right i 32)) buf
+
+let rec write_U64 i buf =
+  let b = Int64.to_int (Int64.logand i 0x7fL) in
+  if 0L <= i && i < 128L then add_byte b buf
+  else (
+    add_byte (b lor 0x80) buf;
+    write_U64 (Int64.shift_right_logical i 7) buf )
+
+let write_U32 i = write_U64 (Int64.logand (Int64.of_int32 i) 0xffffffffL)
+
+let write_U32_of_int i =
+  let i = Int32.of_int i in
+  write_U32 i
+
+let add_bytes str buf =
+  let len = String.length str in
+  write_U32_of_int len buf;
+  add_string str buf
+
+let rec write_S64 i buf =
+  let b = Int64.to_int (Int64.logand i 0x7fL) in
+  if -64L <= i && i < 64L then add_byte b buf
+  else (
+    add_byte (b lor 0x80) buf;
+    write_S64 (Int64.shift_right i 7) buf )
+
+let write_S32 i = write_S64 (Int64.of_int32 i)
+
+let write_F32 f =
+  let i32 = Float32.to_bits f in
+  add_bytes_4 i32
+
+let write_F64 f =
+  let i64 = Float64.to_bits f in
+  add_bytes_8 i64
+
+let indice_encode (idx : binary indice) buf =
+  let (Raw idx) = idx in
+  write_U32_of_int idx buf
+
+let add_char_indice c idx buf =
+  add_char c buf;
+  indice_encode idx buf
+
+let reftype_encode ht buf =
+  match ht with
+  | Func_ht -> add_char '\x70' buf
+  | Extern_ht -> add_char '\x6F' buf
+  | _ -> assert false
+
+let valtype_encode vt buf =
+  let c =
+    match vt with
+    | Num_type I32 -> '\x7F'
+    | Num_type I64 -> '\x7E'
+    | Num_type F32 -> '\x7D'
+    | Num_type F64 -> '\x7C'
+    | Ref_type (Null, Func_ht) -> '\x70'
+    | Ref_type (Null, Extern_ht) -> '\x6F'
+    | _ -> assert false (* vecttype v128 '\x7B' *)
+  in
+  add_char c buf
+
+let vector_encode datas func_encode buf =
+  let vector_buf = Buffer.create 16 in
+  let len =
+    List.fold_left
+      (fun acc data ->
+        match func_encode data vector_buf with
+        | None -> acc
+        | Some () -> acc + 1 )
+      0 datas
+  in
+  write_U32_of_int len buf;
+  Buffer.add_buffer buf vector_buf
+
+let vector_encode_from_namedvalues ~rev_values datas func_encode buf =
+  let { Named.values; _ } = datas in
```

Comment:
```ocaml
let vector_encode_from_namedvalues buf ~rev_values { Named.values; _ } func_encode =
  let { Named.values; _ } = datas in
```

Reply:
Also could you remove `~rev_values` ? It is always true except in the the global case (which looks suspicious), if it's really needed, you can simply do `List.rev` before calling this function.

Reply:
indeed, it's suspicious !? (for `global`), but it seems necessary.
"Hack" because List.rev applies to `values` :
```ocaml
  ....
  let { Named.values; named } = globals in
  let values = List.rev values in
  let globals = { Named.values; named } in
  vector_encode_from_namedvalues globals ...
```